src/librustc_typeck/outlives/implicit_infer.rs

```diff
@@ -244,199 +250,58 @@ fn insert_required_predicates_to_be_wf<'tcx>(
-                    }
-
-                    Component::Projection(proj_ty) => {
-                        // This would arise from something like:
-                        //
-                        // ```
-                        // struct Foo<'a, T: Iterator> {
-                        //    x:  &'a <T as Iterator>::Item
-                        // }
-                        // ```
-                        //
-                        // Here we want to add an explicit `where <T as Iterator>::Item: 'a`.
-                        let ty: Ty<'tcx> = tcx.mk_projection(proj_ty.item_def_id, proj_ty.substs);
-                        required_predicates
-                            .insert(ty::OutlivesPredicate(ty.into(), outlived_region));
-                    }
-
-                    Component::EscapingProjection(_) => {
-                        // As above, but the projection involves
-                        // late-bound regions.  Therefore, the WF
-                        // requirement is not checked in type definition
-                        // but at fn call site, so ignore it.
-                        //
-                        // ```
-                        // struct Foo<'a, T: Iterator> {
-                        //    x: for<'b> fn(<&'b T as Iterator>::Item)
-                        //              //  ^^^^^^^^^^^^^^^^^^^^^^^^^
-                        // }
-                        // ```
-                        //
-                        // Since `'b` is not in scope on `Foo`, can't
-                        // do anything here, ignore it.
-                    }
-
-                    Component::UnresolvedInferenceVariable(_) => bug!("not using infcx"),
+            // ```
+            // trait Trait<'a, T> where Self: 'a  { .. }
+            // ```
+            //
+            // we want to ignore such predicates here, because
+            // there is no type parameter for them to affect. Consider
+            // a struct containing `dyn Trait`:
+            //
+            // ```
+            // struct MyStruct<'x, X> { field: Box<dyn Trait<'x, X>> }
+            // ```
+            //
+            // The `where Self: 'a` predicate refers to the *existential, hidden type*
+            // that is represented by the `dyn Trait`, not to the `X` type parameter
+            // (or any other generic parameter) declared on `MyStruct`.
+            //
+            // Note that we do this check for self **before** applying `substs`. In the
+            // case that `substs` come from a `dyn Trait` type, our caller will have
+            // included `Self = dyn Trait<'x, X>` as the value for `Self`. If we were
+            // to apply the substs, and not filter this predicate, we might then falsely
+            // conclude that e.g. `X: 'x` was a reasonable inferred requirement.
+            if let UnpackedKind::Type(ty) = predicate.0.unpack() {
+                if ty.is_self() && ignore_self_ty {
```

Comment:
@nikomatsakis think we should have a test for `ignore_self_ty` and that it should be work for `dyn` but not for other things.